### ISSUE TITLE
SearXNG sometimes redirects to main page without query

### ISSUE THREAD

**Issue by mooreye** [searxng/searxng#3191]
update from the thread maintainers: Please read the summary in https://github.com/searxng/searxng/issues/3191#issuecomment-2670632164

----

See: https://github.com/libredirect/browser_extension/issues/880

Problem encountered in Firefox extension Libredirect, but they believe this is SearXNG's problem.

**Comment by unixfox** [comment]
It's possible that something does not work great with the anti bot protection.

**Comment by return42** [comment]
Can't reproduce the issue with my instance https://darmarit.org/searx/ or https://paulgo.io/  you mentioned in  https://github.com/libredirect/browser_extension/issues/880#issuecomment-1899420751 .. not sure what the issue is you have.

**Comment by Austin-Olacsi** [comment]
I can say for sure that this is not an issue with libredirect. It has happened to me on some occasions, and i do not have this extension.

i don't know how to reproduce it unfortunately. it just happened to me on my private instance a few moments ago. it usually does not happen. seems very random.

**Comment by gabeklavans** [comment]
Happens to me with enough frequency to be annoying (about once a week). Both when I search using Firefox opensearch (no extensions), and when I use a redirect on iOS safari. It seems to happen with about the same frequency with both access methods. 

The only other variables for me are my DNS provider (cloudflare, on which I have proxy turned off), and nginx being my reverse proxy. But I'm not sure if the problem lies with either of those.

I will try searching directly from the search home page for a while to see if it still occurs.

**Comment by MadAim123** [comment]
I have the same problem. I testes a lot of instances from https://searx.space/ and it happens randomly. I tried it with different browsers and devices and can't resolve it. I don't have any special plugins or extensions installed.

- https://github.com/searxng/searxng/discussions/3219

**Comment by return42** [comment]
@MadAim123 do you have the same issue on my instance? https://darmarit.org/searx/search?q=test&language=de&safesearch=0&categories=general&time_range=month

**Comment by MadAim123** [comment]
@return42 when I tested it now (Firefox, Ege, Chrome on Windows), I can't reproduce it. 
But it happens randomly. So maybe it would happen tomorrow or in 2 weeks - or maybe not. Did you changed something in routing? Or do you use vanilla?

**Comment by return42** [comment]
Vanilla installed by the [scripts](https://docs.searxng.org/admin/installation-scripts.html#installation-scripts) ..

```
./utils/searxng.sh install all
./utils/searxng.sh install apache
```
Not sure its related .. I'm using apache .. AFAIK the docker images are using [caddy](https://github.com/searxng/searxng-docker?tab=readme-ov-file#what-is-included-)

**Comment by MadAim123** [comment]
@return42 I tried it now, after closing my browser after 1 hour and it happend with your instance too.
I klicked on your link (https://darmarit.org/searx/search?q=test&language=de&safesearch=0&categories=general&time_range=month) and was redirected to 
https://darmarit.org/searx/

When I click another time, I was shown the results and I'm not redirected to the startpage. 

With: Firefox, Windows. No extra installed Plugins, no extensions, Cookies are saved, jacascript is enabled.

**Comment by return42** [comment]
Its a pity, I can't reproduce this issue .. 

> Cookies are saved

Did you saved some SearXNG preferences? if so, what prefs .. I am still looking for clues .. may this is related?

**Comment by unixfox** [comment]
@return42 what are the possible ways that the anti bot may force a redirect?

I can only see that it's the case when there are too many requests: https://github.com/searxng/searxng/blob/08e524fc35d382e51725c2778fc97615851b8bc2/searx/botdetection/ip_limit.py#L126

**Comment by return42** [comment]
> what are the possible ways that the anti bot may force a redirect?

good point :+1: / haven't in mind -->

https://github.com/searxng/searxng/blob/c197c0e35e6eaa823d6ba8606df4e5a5c598a07b/searx/botdetection/ip_limit.py#L126

This redirects a browser to the index page / for cases in which (for whatever reasons) the browser has not requested the CSS-ping.

The index page is not in the bot detection and can be loaded even the IP has to many counts in the `ip_limit.SUSPICIOUS_IP_WINDOW` time window.

When the index page is loaded by the browser, the browser will send a CSS-ping request .. 

https://github.com/searxng/searxng/blob/c197c0e35e6eaa823d6ba8606df4e5a5c598a07b/searx/botdetection/ip_limit.py#L113-L118

and the `ip_limit.SUSPICIOUS_IP_WINDOW` time window for this IP will be dropped.

This method is intended to ensure that a normal user is never blocked, even if his IP (for unknown reasons) has had too many accesses in the time window `ip_limit.SUSPICIOUS_IP_WINDOW` the window will be dropped when the index page is loaded.

One reason why a normal user ends up in the time window may be that requests are still coming from the same subnet that do not trigger a CSS ping request. Example: there is a bot and a normal user in the subnet ... then the normal user should not be blocked, even if we have to let the bot pass.

I'll have to analyze this in more detail ... the key is generated here

https://github.com/searxng/searxng/blob/c197c0e35e6eaa823d6ba8606df4e5a5c598a07b/searx/botdetection/link_token.py#L117-L127

And its lifetime is:

https://github.com/searxng/searxng/blob/c197c0e35e6eaa823d6ba8606df4e5a5c598a07b/searx/botdetection/link_token.py#L61-L62

**Comment by mooreye** [comment]
> Did you saved some SearXNG preferences? if so, what prefs .. I am still looking for clues .. may this is related?

For what it's worth, my browser with Libredirect extension has forgetful settings (no cookies, cache, site data saved), all is lost on browser quit and I still experienced the issue on SearXNG instances that I visit for the first time in current session.

**Comment by KrypticKahos** [comment]
If additional information is required for this bug I should be able to assist. I'm having a similar issue, except that it occurs on every search. My setup is as follows:
I have two instances of sear-xng running. One on an unraid machine and the other on a raspberry pi. I have a local reverse proxy setup for failover. 

If I directly connect to the specific searXNG instances I have no issues. Searching works as expected. However if I use the reverse proxy address my searches always redirect to the home page in Firefox. This issue persists on a macbook as well as a windows 11 machine.
Using the exact same setup I don't have any issues using a chrome instance.

**Comment by unixfox** [comment]
If it's personal usage you shouldn't have the anti bot features enabled.

It's not useful.

**Comment by KrypticKahos** [comment]
> If it's personal usage you shouldn't have the anti bot features enabled.
> 
> It's not useful.

I don't believe that is my issue. Don't have any rate limiters enabled that I know of and the config shows it disabled in the docker.

**Comment by Kostrol** [comment]
This is a big, big issue when you are going between a lot of random instances with libredirect daily for each search. 
1. query a new search from the address bar
2. sometimes land on empty page on certain SearXNG instances

Sometimes this will happen multiple times in a row and can get frustrating quickly.


Few selected offending instances I've found so far to be doing this behavior:
https://search.starless.one/
https://s.mble.dk/
https://paulgo.io/
https://search.in.projectsegfau.lt/
https://ooglester.com

**Comment by MimzyFang** [comment]
I'm running into this on my own issue constantly

is there any config I can tweak that makes it less prominent?

**Comment by unixfox** [comment]
It's very easy to replicate, at least on these instances:
- https://search.gcomm.ch/
- https://searx.ankha.ac/
- https://sx.thatxtreme.dev/

Just don't enable javascript and you will get the bug very easily on the first visit.

**Comment by ianweedlun** [comment]
Same deal for me, goes away if I relaunch Firefox. Clearing cache/cookies beforehand doesn't seem to make a difference.

**Comment by gabeklavans** [comment]
I think it's pretty clear this is a confirmed issue. I would suggest anyone considering adding a comment to only add potential solutions or discussions for solutions, so that everyone watching this issue doesn't keep getting "same here" notifications 😁

**Comment by dalf** [comment]
@return42 @unixfox @Bnyro would it make sense to change this line:

https://github.com/searxng/searxng/blob/f195d98bfbcf96e7bf2532419b1f3804ebd739a1/searx/botdetection/ip_limit.py#L125

to something like this:

```python
return flask.redirect(flask.url_for('index', error="too_many_query"), code=302) 
```

the `error` parameter is made up, the point is to let the users who have report issue here to help debugging the issue by looking at the URL.

**Comment by return42** [comment]
To all users affected by this unintentional redirects to SearXNG's main page. A recent discussion on #searxng:matrix.org suggests that this problem may be caused by an inadequate proxy setup.

Make sure that X-Forwarded-For is passed correctly from the proxy, it must be a list of IPs, where the last IP is the proxy's IP and the second last (the first) is the client's IP.  Some proxies transmit an X-Real-IP, this should then be trustworthy / please check whether this is the real IP of the client.

The following error messages in the LOGs suggest an incorrect configuration:

```
WARNING:searx.botdetection: IP from X-Real-IP (random-ip) is not equal to IP from X-Forwarded-For (proxy-ip)
WARNING:searx.botdetection: IP from WSGI environment (proxy-ip) is not equal to IP from X-Real-IP (random-ip)
```

If the client IP is determined incorrectly, for example if the IP of the proxy is mistakenly used instead of the client IP, then the entire bot defense does not work and you are redirected to the main page again and again!

**Comment by return42** [comment]
The client IP is determined by the [get_real_ip(..)](https://docs.searxng.org/src/searx.botdetection.html#searx.botdetection.get_real_ip) function. This function is also used by the "Self information" plugin.

A very simple and minimal check that client IP is determined correctly can be done with this plugin: 

1. first activate the plugin in your preferences ...

![image](https://github.com/searxng/searxng/assets/554536/1a3ba663-14b7-4793-a581-e22d71262cab)

2. Use search term `ip` .. and the IP of your client should be reported on top of the result list ..

![image](https://github.com/searxng/searxng/assets/554536/188ae19c-1b5e-43ea-a320-594d3ff1feca)

The example in the screenshot is a request from my subnet and the IP is the IP of my DSL router.

You can verify the IP by https://whatismyipaddress.com/

**Comment by myned** [comment]
I can confirm that this is related to at least a single proxy setup.

In my case, the `Self Information` plugin was originally returning the IP of my proxy because of a few reasons:
- `X-Forwarded-For: CLIENTIP, PROXYIP` (correct, but `get_real_ip()` returns `PROXYIP` in searxng logs, fixed by adding `x_for = 2` to `limiter.toml`)
- `X-Real-IP: PROXYIP` (incorrect because the searxng-docker Caddyfile uses `header_up X-Real-IP {remote_host}`, fixed by using `{client_ip}` with `trusted_proxies`)
- `WSGI environment: PROXYIP` (incorrect due to https://github.com/searxng/searxng/issues/3632, but did not affect plugin result)

After the above fixes, the plugin now correctly returns the client IP.

**Comment by ac615223s5** [comment]
this is happening not just for redirects but also when searching from the address bar or even visiting the search url directly, also caused by the limiter

**Comment by return42** [comment]
> this is happening not just for redirects but also when searching from the address bar or even visiting the search url directly, also caused by the limiter

Yes, the procedure is the same .. and the solution is the same, see https://github.com/searxng/searxng/issues/3191#issuecomment-2202729527 above.

**Comment by rh-testcode** [comment]
Can someone explain what settings need to be changed for a non-technical user? 

I was able to view my IP as suggested in [#3191(comment)](https://github.com/searxng/searxng/issues/3191#issuecomment-2202729527) (I had to search `ip` because `my ip` didn't return the IP information). The IP returned by SearXNG matched the IP provided by https://whatismyipaddress.com/

I'm using vanilla Firefox on Linux without any redirect add-ons.

**Comment by return42** [comment]
> The IP returned by SearXNG matched the IP provided by https://whatismyipaddress.com/

@rh-testcode and you got redirected from time to time? .. may you have a different issue: is the instance you use maintained by yourself? .. is the instance public? .. can I test the instance online?

**Comment by rh-testcode** [comment]
I get redirected to the main page at least once a day, sometimes more. It happens randomly. I only use public instances. I usually use https://searx.be but if that's not working I'll use https://baresearch.org or another one from the instances list at https://searx.space I use Firefox's built-in default search engine feature.

**Comment by return42** [comment]
@rh-testcode thanks for your feedback .. are you coming over a VPN to the SearXNG instances?

I am still in the analysis phase:  the instances you mentioned are using IPv6 .. perhaps the problem is related to IPv6? (if we assume the redirects are caused by the [rate limit / link_token](https://docs.searxng.org/src/searx.botdetection.html#method-link-token) method).

In the limiter have a [default IPv6 range](https://docs.searxng.org/admin/searx.limiter.html#limiter-toml) of 48 / `ipv6_prefix = 48`.

.. are you coming over a VPN to the SearXNG instances?

My thoughts about (may I'm completely wrong): It could also be that the IP of a VPN access is in the same IP range as a bot that fires the instances ... sounds unlikely, but why else would it only happen once a day?

> I usually use https://searx.be/

@unixfox do you see the same issue on your instance? .. IDK what the cause can be / except my thoughts from above .. any other thoughts or hints for me?

**Comment by return42** [comment]
@rh-testcode thanks for your feedback .. are you coming over a VPN to the SearXNG instances?

I am still in the analysis phase:  the instances you mentioned are using IPv6 .. perhaps the problem is related to IPv6? (if we assume the redirects are caused by the [rate limit / link_token](https://docs.searxng.org/src/searx.botdetection.html#method-link-token) method).

In the limiter have a [default IPv6 range](https://docs.searxng.org/admin/searx.limiter.html#limiter-toml) of 48 / `ipv6_prefix = 48`.

.. are you coming over a VPN to the SearXNG instances?

My thoughts about (may I'm completely wrong): It could also be that the IP of a VPN access is in the same IP range as a bot that fires the instances ... sounds unlikely, but why else would it only happen once a day?

> I usually use https://searx.be/

@unixfox do you see the same issue on your instance? .. IDK what the cause can be / except my thoughts from above .. any other thoughts or hints for me?

**Comment by rh-testcode** [comment]
@return42 I'm not using a VPN.

**Comment by saukkico** [comment]
I have this same issue. Sometimes it happens about once a day, sometimes about once a week. When the redirect happens, I've always done the search with the browser address bar. I've tested with multiple browsers: Brave, Orion, Firefox on mobile and desktop. I don't remember if it happens when searching from the Searxng front page, because I mostly use the address bar for search.

I'm using my own instance which is running behind Traefik on my home server. The instance is also behind a VPN, but I've also tested it without the VPN, and the issue persists. The limiter is enabled. If I search "ip", it gives my actual ip (not the instance ip nor reverse proxy ip), same as provided by https://whatismyipaddress.com/.

**Comment by bearz314** [comment]
In my personal setup, I have a searxng instance (container) served with traefik reverse proxy over Cloudflare CDN/proxy.
`Client --> CF --> Traefik --> Searxng`.

When I first started out, I quickly experienced a constant redirect to homepage and I can't search anything. This is what I found out:

1. Initially my Traefik was passing just the CF IPs to Searxng.
    - This defeats the point of bot protections.
    - It causes false-positive bot-detections, since CF IPs are rotated very quickly.
    - The link token bot protection can break because the HTML and CSS files could be accessed by two different IPs.
    - Fix: Set `forwardedHeaders.trustedIPs` correctly for Traefik to set `X-FORWARDED-FOR = client_IP, CF_IP`
2. Refer number 1, set your `x_for` in `limiter.toml`  appropriately to return the client IP.
    - In my case, `x_for=2` (meaning get the second last). Otherwise, the fast rotating IP is going to cause issues...
    - I get IP mismatch issues but that doesn't cause a problem for me now... (see https://github.com/searxng/searxng/issues/3191#issuecomment-2202729527)
3. Using CF CDN, and I suspect a lot other proxies, heavily caches the content, which means the CSS link token was cached and wasn't fetched.
   - In my tests, the server didn't receive a valid ping from the user. When making a search, the following function is evoked:
https://github.com/searxng/searxng/blob/a1e2b254677a22f1f8968a06564661ac6203c162/searx/botdetection/ip_limit.py#L113
    - This checks if the user has visited the CSS linktoken, and if so, renew the token.
    - On the search result page, the client has a second opportunity to load the linktoken, but it has been cached, it will not, leading to eventual block if the user made several searches in a short time span. This inspired PR #4303.
4. Consider a case where `x_for` is misconfigured and there is caching.
   - You don't do valid pings and your IP is quickly rotating. You get blocked quickly.
5. Consider a better case than number 4: Contents are cached but `x_for` is set correctly.
    - CSS link tokens are not being fetched, at least until fresh link tokens are generated (10 minutes):
https://github.com/searxng/searxng/blob/a1e2b254677a22f1f8968a06564661ac6203c162/searx/botdetection/link_token.py#L57C1-L57C22

    - When making a search, a check is made to see if the user has made a valid ping in the last hour:
https://github.com/searxng/searxng/blob/a1e2b254677a22f1f8968a06564661ac6203c162/searx/botdetection/link_token.py#L60C1-L60C15
    - Because the client IP is correct, the token is renewed successfully.

Key takeaways: Check caching policies and check if configurations allow client IPs to pass in correctly.

**Comment by return42** [comment]
Should be fixed by 

- https://github.com/searxng/searxng/pull/4303 ..

If you continue to observe this problem, please note that this error is usually caused by a faulty setup of the network infrastructure and is usually not a real bug in SearXNG 

Before post to this thread, first please follow the recommendations from this comments:

- https://github.com/searxng/searxng/issues/3191#issuecomment-2202729527
- https://github.com/searxng/searxng/issues/3191#issuecomment-2202808807
- https://github.com/searxng/searxng/issues/3191#issuecomment-2648536441

**Comment by pesschap** [comment]
I do not have the bot detection on, no redis, and I am experiencing this issue on the latest release.

**Comment by return42** [comment]
> I do not have the bot detection on, no redis, and I am experiencing this issue on the latest release.

If it is in the last commits, then I think its related to:

- https://github.com/searxng/searxng/pull/4448

**Comment by rhsCZ** [comment]
I would like to discuss issue from #4621. I have looked to logs and i see that it may be related with botdetection.
I have like almost 15k entries in 1 log file of bot detection with different ip's (some IP's repeats in different time).
If i will turn off bot detection it maybe hurt the instance.
There is possibility, that many client's are behind NAT from ISP with same shared IP. Could it lead to be blocked by bot detection?
Any ideas?
P.S. no Cloudflare DNS or CDN are involed.

**Comment by suazojaime** [comment]
I found a way to go around this problem by rolling back to this image.
image: docker.io/searxng/searxng:2025.4.17-777ba6ddf

After that I dont get redirected to the main page anymore. I have try many fixes described here, but the rollback its the only thing that have worked to me.

**Comment by unixfox** [comment]
Hello @suazojaime, could you join the matrix channel of searxng in order to discuss your issue: https://matrix.to/#/#searxng:matrix.org
